File: solution/kotlin/day17/src/main/kotlin/games/FizzBuzz.kt
package games

import arrow.core.None
import arrow.core.Option
import arrow.core.Some

const val MIN = 1
const val MAX = 100

object FizzBuzz {
    fun convert(input: Int): Option<String> = when {
        isOutOfRange(input) -> None
        else -> Some(convertSafely(input))
    }

    private fun convertSafely(input: Int): String = when {
        `is`(15, input) -> "FizzBuzz"
        `is`(3, input) -> "Fizz"
        `is`(5, input) -> "Buzz"
        else -> input.toString()
    }

    private fun `is`(divisor: Int, input: Int): Boolean = input % divisor == 0
    private fun isOutOfRange(input: Int) = input < MIN || input > MAX
}

File: exercise/ts/day10/src/fizzbuzz.ts
const min = 0;
const max = 100;
const fizz = 3;
const buzz = 5;
const fizzBuzz = 15;

export function fizzbuzz(input: number): string | number {
    if (isOutOfRange(input)) {
        throw new Error("Input is out of range")
    }
    return convertSafely(input);
}

function convertSafely(input: number): string | number {
    if (is(fizzBuzz, input)) {
        return "FizzBuzz";
    }
    if (is(fizz, input)) {
        return "Fizz";
    }
    if (is(buzz, input)) {
        return "Buzz";
    }
    return input;
}

function is(divisor: number, input: number): boolean {
    return input % divisor === 0;
}

function isOutOfRange(input: number): boolean {
    return input <= min || input > max;
}